File: skills/test-driven-development/scripts/validate_tdd_evidence.py
#!/usr/bin/env python3
"""Validate Markdown reports for TDD evidence or an explicit TDD exception."""

from __future__ import annotations

import argparse
import json
import re
import sys
from dataclasses import asdict
from dataclasses import dataclass
from pathlib import Path


EVIDENCE_HEADING_RE = re.compile(r"^\s{0,3}#{1,6}\s+TDD 证据\s*$", re.IGNORECASE | re.MULTILINE)
EXCEPTION_HEADING_RE = re.compile(
    r"^\s{0,3}#{1,6}\s+TDD 例外记录\s*$", re.IGNORECASE | re.MULTILINE
)
SPEC_SOURCE_RE = re.compile(
    r"\b(?:REQ|API|SCHEMA|STATE|ERR|AC|NFR|MIG|OBS|NON)(?:-[A-Z0-9]+)*-\d{3,}\b|bug|复现|验收|acceptance",
    re.IGNORECASE,
)
PLACEHOLDER_RE = re.compile(r"<[^>\n]{2,}>|\[[^\]\n]{2,}\]|\bTODO\b|\bTBD\b|待补充|待定|占位|\.\.\.")
SUSPICIOUS_RE = re.compile(
    r"先实现|后补测|补测试|测试后补|直接实现|先写代码|实现后写测试|test after|implemented first",
    re.IGNORECASE,
)

EVIDENCE_FIELDS = (
    "规格/缺陷/验收",
    "RED 测试",
    "RED 命令",
    "RED 失败",
    "GREEN 变更",
    "GREEN 命令",
    "REFACTOR 命令",
    "最终验证",
)
OPTIONAL_TEST_TYPE_FIELD = "测试类型"
ALLOWED_TEST_TYPES = {"behavior", "contract", "architecture", "source-scan", "config"}
BEHAVIOR_SOURCE_RE = re.compile(r"用户|点击|按钮|页面|界面|交互|流程|UI|widget|behavior", re.IGNORECASE)
EXCEPTION_FIELDS = (
    "原因",
    "用户批准",
    "替代验证",
    "风险",
)


@dataclass(frozen=True)
class ValidationResult:
    path: str
    ok: bool
    errors: list[str]
    warnings: list[str]


def parse_args(argv: list[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Validate Markdown TDD evidence reports for RED/GREEN/REFACTOR proof."
    )
    parser.add_argument("evidence_files", nargs="+", help="Path(s) to Markdown report files.")
    parser.add_argument(
        "--format",
        choices=("text", "json"),
        default="text",
        help="Output format. Use json for CI or editor integrations.",
    )
    parser.add_argument(
        "--strict",
        action="store_true",
        help="Treat warning-level weak evidence as validation errors.",
    )
    return parser.parse_args(argv)


def field_pattern(label: str) -> re.Pattern[str]:
    escaped = re.escape(label)
    return re.compile(
        rf"^\s*(?:[-*]\s*)?(?:\*\*)?{escaped}\s*(?::\*\*|\*\*\s*:|:)\s*(.+?)\s*$",
        re.IGNORECASE | re.MULTILINE,
    )


def get_field(text: str, label: str) -> str | None:
    match = field_pattern(label).search(text)
    if not match:
        return None
    return match.group(1).strip()


def is_placeholder(value: str) -> bool:
    cleaned = value.strip().strip("`").strip()
    return not cleaned or bool(PLACEHOLDER_RE.search(cleaned))


def require_fields(text: str, labels: tuple[str, ...], section_name: str) -> list[str]:
    errors: list[str] = []
    for label in labels:
        value = get_field(text, label)
        if value is None:
            errors.append(f"{section_name} is missing required field: {label}.")
        elif is_placeholder(value):
            errors.append(f"{section_name} field {label!r} still contains a placeholder: {value!r}.")
    return errors


def validate_text(text: str, strict: bool) -> tuple[list[str], list[str]]:
    errors: list[str] = []
    warnings: list[str] = []

    if not text.strip():
        return ["Evidence file is empty."], warnings

    if SUSPICIOUS_RE.search(text):
        errors.append("Suspicious after-the-fact testing wording found; TDD requires RED before implementation.")

    has_evidence = bool(EVIDENCE_HEADING_RE.search(text))
    has_exception = bool(EXCEPTION_HEADING_RE.search(text))

    if not has_evidence and not has_exception:
        errors.append("Missing TDD 证据 or TDD 例外记录 section.")

    if has_evidence:
        errors.extend(require_fields(text, EVIDENCE_FIELDS, "TDD 证据"))
        source = get_field(text, "规格/缺陷/验收")
        test_type = get_field(text, OPTIONAL_TEST_TYPE_FIELD)
        red_failure = get_field(text, "RED 失败")
        final_verification = get_field(text, "最终验证")

        if source and not is_placeholder(source) and not SPEC_SOURCE_RE.search(source):
            warnings.append(
                "规格/缺陷/验收 does not look traceable to a Spec ID, bug reproduction, or acceptance criterion."
            )

        if test_type:
            normalized_test_type = test_type.strip().strip("`").lower()
            if normalized_test_type not in ALLOWED_TEST_TYPES:
                errors.append(
                    "invalid 测试类型: "
                    + test_type
                    + ". Allowed: "
                    + ", ".join(sorted(ALLOWED_TEST_TYPES))
                    + "."
                )
            elif normalized_test_type == "source-scan" and source and BEHAVIOR_SOURCE_RE.search(source):
                warnings.append("source-scan should not be used as the primary proof for user-facing behavior.")

        if red_failure and re.search(r"\bpass(?:ed)?\b|通过", red_failure, re.IGNORECASE):
            warnings.append("RED 失败提到了类似通过的结果；请确认测试确实先失败。")

        if final_verification and re.search(r"not run|未运行|无法运行", final_verification, re.IGNORECASE):
            warnings.append("最终验证未运行；如果自动化被阻塞，请使用 TDD 例外记录。")

    if has_exception:
        errors.extend(require_fields(text, EXCEPTION_FIELDS, "TDD 例外记录"))

    if strict and warnings:
        errors.extend(warnings)
        warnings = []

    return errors, warnings


def build_result(path: Path, strict: bool) -> ValidationResult:
    if not path.exists():
        return ValidationResult(path=str(path), ok=False, errors=[f"File does not exist: {path}"], warnings=[])
    if not path.is_file():
        return ValidationResult(path=str(path), ok=False, errors=[f"Path is not a file: {path}"], warnings=[])

    text = path.read_text(encoding="utf-8")
    errors, warnings = validate_text(text, strict)
    return ValidationResult(path=str(path), ok=not errors, errors=errors, warnings=warnings)


def print_text_results(results: list[ValidationResult]) -> None:
    for index, result in enumerate(results):
        if index:
            print()

        if result.errors:
            print(f"TDD evidence validation failed: {result.path}")
            print("\nErrors:")
            for error in result.errors:
                print(f"- {error}")
        else:
            print(f"TDD evidence validation passed: {result.path}")

        if result.warnings:
            print("\nWarnings:")
            for warning in result.warnings:
                print(f"- {warning}")


def main(argv: list[str]) -> int:
    args = parse_args(argv)
    results = [build_result(Path(evidence_file), args.strict) for evidence_file in args.evidence_files]
    ok = all(result.ok for result in results)

    if args.format == "json":
        payload = {
            "ok": ok,
            "error_count": sum(len(result.errors) for result in results),
            "warning_count": sum(len(result.warnings) for result in results),
            "results": [asdict(result) for result in results],
        }
        print(json.dumps(payload, ensure_ascii=False, indent=2))
    else:
        print_text_results(results)

    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: src/lib/documents/validate-tdd-evidence.ts
import { existsSync, readFileSync, statSync } from "node:fs";
import { dirname, resolve } from "node:path";

import { isPathIgnoredByGitignore } from "./artifact-mode.ts";
import type { ArtifactModeValue } from "./artifact-mode.ts";

const EVIDENCE_HEADING_RE = /^\s{0,3}#{1,6}\s+TDD 证据\s*$/gim;
const EXCEPTION_HEADING_RE = /^\s{0,3}#{1,6}\s+TDD 例外记录\s*$/gim;
const SPEC_SOURCE_RE = /\b(?:REQ|API|SCHEMA|STATE|ERR|AC|NFR|MIG|OBS|NON)(?:-[A-Z0-9]+)*-\d{3,}\b|bug|复现|验收|acceptance/i;
const PLACEHOLDER_RE = /<[^>\n]{2,}>|\[[^\]\n]{2,}\]|\bTODO\b|\bTBD\b|待补充|待定|占位|\.\.\./;
const SUSPICIOUS_RE = /先实现|后补测|补测试|测试后补|直接实现|先写代码|实现后写测试|test after|implemented first/i;
const BEHAVIOR_SOURCE_RE = /用户|点击|按钮|页面|界面|交互|流程|UI|widget|behavior/i;

const EVIDENCE_FIELDS = [
  "规格/缺陷/验收",
  "RED 测试",
  "RED 命令",
  "RED 失败",
  "GREEN 变更",
  "GREEN 命令",
  "REFACTOR 命令",
  "最终验证",
];
const REPRODUCIBLE_REFERENCE_FIELDS = ["RED 测试", "RED 命令", "GREEN 命令", "REFACTOR 命令", "最终验证"];
const OPTIONAL_TEST_TYPE_FIELD = "测试类型";
const ALLOWED_TEST_TYPES = new Set(["behavior", "contract", "architecture", "source-scan", "config"]);
const EXCEPTION_FIELDS = ["原因", "用户批准", "替代验证", "风险"];

export interface ValidationResult {
  path: string;
  ok: boolean;
  content_valid: boolean;
  valid_for: Array<"local-review" | "task-completion" | "formal-completion">;
  formal_completion_allowed: boolean;
  errors: string[];
  warnings: string[];
}

export interface ValidationPayload {
  ok: boolean;
  error_count: number;
  warning_count: number;
  results: ValidationResult[];
}

export interface ValidationOptions {
  strict: boolean;
  root?: string;
  artifactMode?: ArtifactModeValue;
}

type ValidationInput = boolean | ValidationOptions;

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function fieldPattern(label: string): RegExp {
  return new RegExp(
    `^\\s*(?:[-*]\\s*)?(?:\\*\\*)?${escapeRegExp(label)}\\s*(?::\\*\\*|\\*\\*\\s*:|:)\\s*(.+?)\\s*$`,
    "gim",
  );
}

function getField(text: string, label: string): string | undefined {
  const match = fieldPattern(label).exec(text);
  return match ? match[1].trim() : undefined;
}

function getFields(text: string, label: string): string[] {
  return [...text.matchAll(fieldPattern(label))].map((match) => match[1].trim());
}

function isPlaceholder(value: string): boolean {
  const cleaned = value.trim().replace(/^`+|`+$/g, "").trim();
  return !cleaned || PLACEHOLDER_RE.test(cleaned);
}

function requireFields(text: string, labels: string[], sectionName: string): string[] {
  const errors: string[] = [];
  for (const label of labels) {
    const value = getField(text, label);
    if (value === undefined) {
      errors.push(`${sectionName} is missing required field: ${label}.`);
    } else if (isPlaceholder(value)) {
      errors.push(`${sectionName} field '${label}' still contains a placeholder: '${value}'.`);
    }
  }
  return errors;
}

function normalizeOptions(input: ValidationInput): ValidationOptions {
  if (typeof input === "boolean") {
    return { strict: input };
  }
  return input;
}

function localPathReferences(text: string): string[] {
  const withoutQuotedSearchPatterns = text.replace(/"[^"\n]*"|'[^'\n]*'/g, " ");
  return [...new Set([...withoutQuotedSearchPatterns.matchAll(/\b(?:tests|src|skills|docs|hooks|assets|\.github)\/[A-Za-z0-9_.\/-]+/g)]
    .map((match) => match[0])
    .filter((reference) => !reference.includes("...") && /\.[A-Za-z0-9]+$/.test(reference)))]
    .sort();
}

function referencedPathExists(root: string, evidencePath: string, reference: string): boolean {
  const candidates = [resolve(root, reference), resolve(dirname(evidencePath), reference)];
  return candidates.some((candidate) => existsSync(candidate));
}

function validateEvidenceContext(path: string, text: string, options: ValidationOptions): string[] {
  const errors: string[] = [];
  if (!options.root) {
    return errors;
  }

  const root = resolve(options.root);
  const evidencePath = resolve(path);
  if (options.artifactMode !== "local" && isPathIgnoredByGitignore(root, evidencePath)) {
    errors.push("ignored evidence cannot be used as formal completion evidence.");
  }

  const reproducibleText = REPRODUCIBLE_REFERENCE_FIELDS.flatMap((field) => getFields(text, field)).join("\n");
  for (const reference of localPathReferences(reproducibleText)) {
    if (!referencedPathExists(root, evidencePath, reference)) {
      errors.push(`referenced path does not exist: ${reference}`);
    }
  }
  return errors;
}

export function validateText(text: string, input: ValidationInput): [string[], string[]] {
  const options = normalizeOptions(input);
  const errors: string[] = [];
  let warnings: string[] = [];

  if (!text.trim()) {
    return [["Evidence file is empty."], warnings];
  }

  if (SUSPICIOUS_RE.test(text)) {
    errors.push("Suspicious after-the-fact testing wording found; TDD requires RED before implementation.");
  }

  const hasEvidence = EVIDENCE_HEADING_RE.test(text);
  EVIDENCE_HEADING_RE.lastIndex = 0;
  const hasException = EXCEPTION_HEADING_RE.test(text);
  EXCEPTION_HEADING_RE.lastIndex = 0;

  if (!hasEvidence && !hasException) {
    errors.push("Missing TDD 证据 or TDD 例外记录 section.");
  }

  if (hasEvidence) {
    errors.push(...requireFields(text, EVIDENCE_FIELDS, "TDD 证据"));
    const source = getField(text, "规格/缺陷/验收");
    const testType = getField(text, OPTIONAL_TEST_TYPE_FIELD);
    const redFailure = getField(text, "RED 失败");
    const finalVerification = getField(text, "最终验证");

    if (source && !isPlaceholder(source) && options.artifactMode !== "local" && !SPEC_SOURCE_RE.test(source)) {
      warnings.push("规格/缺陷/验收 does not look traceable to a Spec ID, bug reproduction, or acceptance criterion.");
    }

    if (testType) {
      const normalizedTestType = testType.trim().replace(/^`+|`+$/g, "").toLowerCase();
      if (!ALLOWED_TEST_TYPES.has(normalizedTestType)) {
        errors.push(`invalid 测试类型: ${testType}. Allowed: ${[...ALLOWED_TEST_TYPES].sort().join(", ")}.`);
      } else if (normalizedTestType === "source-scan" && source && BEHAVIOR_SOURCE_RE.test(source)) {
        warnings.push("source-scan should not be used as the primary proof for user-facing behavior.");
      }
    }

    if (redFailure && /\bpass(?:ed)?\b|通过/i.test(redFailure)) {
      warnings.push("RED 失败提到了类似通过的结果；请确认测试确实先失败。");
    }

    if (finalVerification && /not run|未运行|无法运行/i.test(finalVerification)) {
      warnings.push("最终验证未运行；如果自动化被阻塞，请使用 TDD 例外记录。");
    }
  }

  if (hasException) {
    errors.push(...requireFields(text, EXCEPTION_FIELDS, "TDD 例外记录"));
  }

  if (options.strict && warnings.length > 0) {
    errors.push(...warnings);
    warnings = [];
  }

  return [errors, warnings];
}

export function buildResult(path: string, input: ValidationInput): ValidationResult {
  const options = normalizeOptions(input);
  if (!existsSync(path)) {
    return { path, ok: false, content_valid: false, valid_for: [], formal_completion_allowed: false, errors: [`File does not exist: ${path}`], warnings: [] };
  }
  if (!statSync(path).isFile()) {
    return { path, ok: false, content_valid: false, valid_for: [], formal_completion_allowed: false, errors: [`Path is not a file: ${path}`], warnings: [] };
  }
  const text = readFileSync(path, "utf8");
  const [contentErrors, warnings] = validateText(text, options);
  const contextErrors = validateEvidenceContext(path, text, options);
  const errors = [...contentErrors, ...contextErrors];
  const contentValid = contentErrors.length === 0;
  const ignored = options.root ? isPathIgnoredByGitignore(resolve(options.root), resolve(path)) : false;
  const formalCompletionAllowed = contentValid && contextErrors.length === 0 && options.artifactMode !== "local" && !ignored;
  return {
    path,
    ok: errors.length === 0,
    content_valid: contentValid,
    valid_for: contentValid
      ? formalCompletionAllowed ? ["local-review", "task-completion", "formal-completion"] : ["local-review"]
      : [],
    formal_completion_allowed: formalCompletionAllowed,
    errors,
    warnings,
  };
}

export function buildPayload(paths: string[], input: ValidationInput): ValidationPayload {
  const options = normalizeOptions(input);
  const results = paths.map((path) => buildResult(path, options));
  return {
    ok: results.every((result) => result.ok),
    error_count: results.reduce((total, result) => total + result.errors.length, 0),
    warning_count: results.reduce((total, result) => total + result.warnings.length, 0),
    results,
  };
}

export function formatTextResults(results: ValidationResult[]): string {
  return results
    .map((result) => {
      const lines: string[] = [];
      if (result.errors.length > 0) {
        lines.push(`TDD evidence validation failed: ${result.path}`, "", "Errors:");
        for (const error of result.errors) {
          lines.push(`- ${error}`);
        }
      } else {
        lines.push(`TDD evidence validation passed: ${result.path}`);
      }
      if (result.warnings.length > 0) {
        lines.push("", "Warnings:");
        for (const warning of result.warnings) {
          lines.push(`- ${warning}`);
        }
      }
      return lines.join("\n");
    })
    .join("\n\n");
}
